native/jni-native.targets

```
@@ -0,0 +1,201 @@
+      <SourceFiles>@(NativeSource-> '&quot;%(rootdir)%(directory)%(filename)%(extension)&quot;', ' ')</SourceFiles>
+    </PropertyGroup>
+
+    <PropertyGroup>
+      <CompilerArgs>/O2 /D WIN32 /D NDEBUG /D _WINDLL /D _MBCS /GS /W3 /nologo</CompilerArgs>
+      <LinkerArgs>/DLL user32.lib $(NativeName).res /out:&quot;$(CompiledResult)&quot;</LinkerArgs>
+    </PropertyGroup>
+    <PropertyGroup Condition="'$(NativeConfiguration)'=='Debug'">
+      <CompilerArgs>/Od /D WIN32 /D _WINDLL /D _MBCS /GS /W3 /nologo</CompilerArgs>
+      <LinkerArgs>/DLL /DEBUG user32.lib $(NativeName).res /out:&quot;$(CompiledResult)&quot;</LinkerArgs>
+    </PropertyGroup>
+
+    <MakeDir Directories="$(NativeObjDir)" />
+
+    <ItemGroup>
+      <Tokens Include="VERSIONLIST">
+        <ReplacementValue>$(Version.Replace('.',','))</ReplacementValue>
+      </Tokens>
+      <Tokens Include="VERSION">
+        <ReplacementValue>$(Version)</ReplacementValue>
+      </Tokens>
+      <Tokens Include="ARCH">
+        <ReplacementValue>$(NativePlatform)</ReplacementValue>
+      </Tokens>
+    </ItemGroup>
+
+    <!-- Generate the RC file -->
+    <TemplateFile
+        Template="$(NativeSourceDir)$(NativeName).rc.in"
+        OutputFileName="$(NativeObjDir)$(NativeName).rc"
+        Tokens="@(Tokens)" />
+
+    <!-- Generate the resource file -->
+    <Exec Command="&quot;$(MSVCWinSDKToolPath)\rc.exe&quot; /nologo &quot;$(NativeObjDir)$(NativeName).rc&quot;"
+        WorkingDirectory="$(NativeObjDir)"
+        ConsoleToMSBuild="true" />
+
+    <!-- Compile ikvm-native binary -->
+    <Exec Command="&quot;$(MSVCCompilerPath)&quot; $(IncPaths) $(CompilerArgs) $(SourceFiles) /link $(LibPaths) $(LinkerArgs)"
+        WorkingDirectory="$(NativeObjDir)"
+        ConsoleToMSBuild="true" />
+
+    <!-- Copy the results to the project output directory -->
+    <Copy
+        SourceFiles="$(CompiledResult)"
+        DestinationFiles="@(NativeBinary)" />
+  </Target>
+
+  <ItemGroup>
+    <PackageReference Include="CompilerPaths" Version="1.0.1"
+      Condition="$([MSBuild]::IsOsPlatform(Windows))" />
+  </ItemGroup>
+
+  <!--
+    Target used to build the IKVM native binary on *nix or macOS
+  -->
+  <Target
+    Name="Build-IKVM-Native-Unix"
+    AfterTargets="CoreCompile"
+    Inputs="@(NativeSource)"
+    Outputs="@(NativeBinary)"
+    Condition="$([MSBuild]::IsOsPlatform(Linux)) OR $([MSBuild]::IsOsPlatform(OSX))">
+
+    <!-- gcc compiler flags -->
+    <PropertyGroup>
+      <CompiledResult>$(NativeObjDir)$(NativeBinaryName)</CompiledResult>
+      <CompilerArgs>--shared -fPIC</CompilerArgs>
+      <SourceFiles>@(NativeSource-> '&quot;%(rootdir)%(directory)%(filename)%(extension)&quot;', ' ')</SourceFiles>
+    </PropertyGroup>
+
+    <MakeDir Directories="$(NativeObjDir)" />
+
+    <!-- Compile ikvm-native binary -->
+    <Exec Command="&quot;gcc&quot; -o &quot;$(CompiledResult)&quot; $(CompilerArgs) `pkg-config --cflags --libs gmodule-2.0` $(SourceFiles)"
```

```suggestion
    <Exec Command="&quot;gcc&quot; -o &quot;$(CompiledResult)&quot; $(CompilerArgs) `pkg-config --cflags gmodule-2.0` $(SourceFiles) `pkg-config --libs gmodule-2.0`"
```

I found that the original command line structure worked fine on OSX, where gcc is an alias for clang/LLVM. On Ubuntu with a "real" gcc  installation, the compiler appeared to be sensitive to the order of the flags, and the resulting library would not load when compiled with the original configuration as it couldn't find the library dependencies. Ensuring that the linker flags come last seemed to fix the issue, as per the suggested change above.